### PR TITLE
Add Seeding, MaxStepReached, and Bootstrapping fix

### PR DESCRIPTION
**Pull request by awjuliani**
* Add ability to seed learning (numpy, tensorflow, and Unity) with `--seed` flag.
* Add `maxStepReached` flag to Agents and Academy.
* Change way value bootstrapping works in PPO to take advantage of timeouts.
* Default size of GridWorld changed to 5x5 in order to validate bootstrapping changes.

### CODE REVIEW

File: python/learn.py [hunk 1]
```diff
@@ -71,7 +73,11 @@ def get_progress():
     lesson = int(options['--lesson'])
     fast_simulation = not bool(options['--slow'])
 
-    env = UnityEnvironment(file_name=env_name, worker_id=worker_id, curriculum=curriculum_file)
+    if seed != -1:
```
Comment (eshvk) [inline python/learn.py]: Can you explain a bit why this check exists? Why is the default pseudo randomizer explicitly to -1?

File: python/learn.py [hunk 2]
```diff
@@ -43,12 +43,13 @@ def get_progress():
     Options:
       --help                     Show this message.
       --curriculum=<file>        Curriculum json file for environment [default: None].
-      --slow                     Whether to run the game at training speed [default: False].
       --keep-checkpoints=<n>     How many model checkpoints to keep [default: 5].
       --lesson=<n>               Start learning from this lesson [default: 0].
       --load                     Whether to load the model or randomly initialize [default: False].
-      --run-path=<path>          The sub-directory name for model and summary statistics [default: ppo]. 
+      --run-id=<path>            The sub-directory name for model and summary statistics [default: ppo].
```
Comment (eshvk) [inline python/learn.py]: If this is branching from dev-0.3, I think this is a typo and should be --run-path. model_path below uses --run-path anyways.

File: python/trainer_configurations.yaml [hunk 3]
```diff
@@ -1,27 +1,40 @@
 default:
-    batch_size: 256
-    beta: 2.5e-3
-    buffer_size: 5000
+    batch_size: 32
+    beta: 5.0e-3
```
Comment (eshvk) [inline python/trainer_configurations.yaml]: Would prefer adding a single line comment articulating the change in default parameters. Perhaps referencing a trial run?

File: python/trainers/ppo_trainer.py [hunk 4]
```diff
@@ -46,15 +46,16 @@ def __init__(self, sess, env, brain_name, trainer_parameters, training):
...
                 raise UnityTrainerException("The memory size for brain {0} is 0 even though the trainer uses recurrent."
                                             .format(brain_name))
-            elif (self.m_size % 4 != 0):
+            elif self.m_size % 4 != 0:
                 raise UnityTrainerException("The memory size for brain {0} is {1} but it must be divisible by 4."
                                             .format(brain_name, self.m_size))
 
         self.variable_scope = trainer_parameters['graph_scope']
         with tf.variable_scope(self.variable_scope):
+            tf.set_random_seed(seed)
```
Comment (eshvk) [inline python/trainers/ppo_trainer.py]: So in the master `learn.py` you check if `seed != -1` before setting the seed for `tf`. However PPOTrainer is again called with the `seed` which may potentially be `-1` which means that here we could end up setting `tf.set_random_seed(-1)` ?

File: unity-environment/Assets/ML-Agents/Scripts/ExternalCommunicator.cs [hunk 5]
```diff
@@ -30,9 +30,11 @@ public class ExternalCommunicator : Communicator
     List<float> concatenatedRewards = new List<float>(32);
     List<float> concatenatedMemories = new List<float>(1024);
     List<bool> concatenatedDones = new List<bool>(32);
+    List<bool> concatenatedMaxes = new List<bool>(32);
```
Comment (eshvk) [inline unity-environment/Assets/ML-Agents/Scripts/ExternalCommunicator.cs]: This are set as sane defaults. Perhaps we can do a `public const int DEFAULT_NUM_AGENTS = 32` before hand?

File: python/trainers/ppo_trainer.py [hunk 6]
```diff
@@ -246,9 +246,8 @@ def process_experiences(self, info):
         for l in range(len(info.agents)):
             agent_actions = self.training_buffer[info.agents[l]]['actions']
             if ((info.local_done[l] or len(agent_actions) > self.trainer_parameters['time_horizon'])
```
Comment (vincentpierre) [inline python/trainers/ppo_trainer.py]: Don't you need to check if `info.max_reached[l]` in this line ?

File: python/trainers/imitation_trainer.py [hunk 7]
```diff
@@ -44,7 +45,7 @@ def __init__(self, state_size, action_size, h_size, lr, action_type, n_layers):
 
 class ImitationTrainer(Trainer):
     """The ImitationTrainer is an implementation of the imitation learning."""
-    def __init__(self, sess, env, brain_name, trainer_parameters, training):
+    def __init__(self, sess, env, brain_name, trainer_parameters, training, seed):
         """
```
Comment (vincentpierre) [inline python/trainers/imitation_trainer.py]: You need to make use of this seed.

File: python/unityagents/environment.py [hunk 8]
```diff
@@ -90,7 +91,8 @@ def __init__(self, file_name, worker_id=0,
             # Launch Unity environment
             proc1 = subprocess.Popen(
                 [launch_string,
-                 '--port', str(self.port)])
+                 '--port', str(self.port),
+                 '--seed', str(seed)])
```
Comment (vincentpierre) [inline python/unityagents/environment.py]: Do not send the seed if the seed is -1 or None
